lua/plenary/git.lua

```diff
@@ -0,0 +1,146 @@
+local Job = require 'plenary.job'
+
+local git = {}
+
+-- TODO(conni2461): What else? Need more ideas :)
+
+--- Creates a git commit table
+--- Requires format: --format=%h:%an:%aD:%s
+git.make_commit = function(commit_line)
+  local rev, author, date, msg = string.match(commit_line, '([^:]+):([^:]+):([^:]+):(.*)')
+  return {
+    revision = rev,
+    author = author,
+    date = date,
+    message = mes,
+  }
+end
+
+git.on_commits = function(opts, each_callback, exit_callback)
+  opts = opts or {}
+
+  opts.cwd = opts.cwd or vim.loop.cwd()
+  opts.file = opts.file or nil
+
+  local output = {}
+  local j = Job:new {
+    command = 'git',
+    args = { '--no-pager', 'log', '--format=%h:%an:%aD:%s', file },
+    cwd = cwd,
+    on_stdout = function(_, line)
+      local commit = git.make_commit(line)
+      table.insert(output, commit)
+      if each_callback then
+        each_callback(commit)
+      end
+    end,
+    on_exit = function()
+      if exit_callback then
+        exit_callback(output)
+      end
+    end
+  }
+
+  if not each_callback and not exit_callback then
+    j:sync()
+    return output
+  else
+    j:start()
+  end
+end
+
+
+-- TODO(conni2461): Need to be reworked to work like git.on_commits
+git.tags = function(cwd, callback)
+  local output = {}
+  Job:new({
+    command = 'git',
+    args = { 'tag' },
+    cwd = cwd,
+    on_stdout = function(_, line)
+      table.insert(output, line)
+    end,
+    on_exit = function()
+      callback(output)
+    end
+  }):start()
+end
+
+git.branches = function(cwd, callback)
+  local output = {}
+  Job:new({
+    command = 'git',
+    args = { 'branch', '--all' },
+    cwd = cwd,
+    on_stdout = function(_, line)
+      if not string.find(line, 'HEAD') then
```

Comment:
Why HEAD is important to be found here?

Reply:
That is just copied from telescope. There should be just one head. `remotes/origin/HEAD -> origin/master` and that would be a duplication for `remotes/origin/master`

Reply:
Ohhh i see thanks